agent/testagent.go

```diff
@@ -123,79 +131,109 @@ func (a *TestAgent) Start(t *testing.T) *TestAgent {
+			path := filepath.Join(a.Config.DataDir, filename)
+			if err := initKeyring(path, key); err != nil {
+				return fmt.Errorf("Error creating keyring %s: %s", path, err)
 			}
-			writeKey(a.Key, SerfLANKeyring)
-			writeKey(a.Key, SerfWANKeyring)
+			return nil
 		}
-
-		logOutput := a.LogOutput
-		if logOutput == nil {
-			logOutput = os.Stderr
+		err = writeKey(a.Key, SerfLANKeyring)
+		if err != nil {
+			return err
+		}
+		err = writeKey(a.Key, SerfWANKeyring)
+		if err != nil {
+			return err
 		}
-		agentLogger := log.New(logOutput, a.Name+" - ", log.LstdFlags|log.Lmicroseconds)
-
-		agent, err := New(a.Config, agentLogger)
-		require.NoError(err, fmt.Sprintf("Error creating agent: %s", err))
-
-		agent.LogOutput = logOutput
-		agent.LogWriter = a.LogWriter
-		agent.MemSink = metrics.NewInmemSink(1*time.Second, time.Minute)
-
-		// we need the err var in the next exit condition
-		if err := agent.Start(); err == nil {
-			a.Agent = agent
-			break
-		} else if i == 0 {
-			require.Fail("%s %s Error starting agent: %s", id, a.Name, err)
-		} else if a.ExpectConfigError {
+	}
+
+	logOutput := a.LogOutput
+	if logOutput == nil {
+		logOutput = os.Stderr
+	}
+	agentLogger := log.New(logOutput, a.Name+" - ", log.LstdFlags|log.Lmicroseconds)
+
+	agent, err := New(a.Config, agentLogger)
+	if err != nil {
+		return fmt.Errorf("Error creating agent: %s", err)
+	}
+
+	agent.LogOutput = logOutput
+	agent.LogWriter = a.LogWriter
+	agent.MemSink = metrics.NewInmemSink(1*time.Second, time.Minute)
+
+	if err := agent.Start(); err != nil {
+		agent.ShutdownAgent()
+		agent.ShutdownEndpoints()
+
+		if a.ExpectConfigError {
 			// Panic the error since this can be caught if needed. Pretty gross way to
 			// detect errors but enough for now and this is a tiny edge case that I'd
 			// otherwise not have a way to test at all...
+			//
+			// TODO(sadams): This can be refactored away by returning an
+			// error here instead of panicing, removing the `ExpectConfigError`
+			// field from `TestAgent`, and having the test that uses this
+			// (TestAgent_ConnectClusterIDConfig) check for an error instead of
+			// catching a panic.
 			panic(err)
-		} else {
-			agent.ShutdownAgent()
-			agent.ShutdownEndpoints()
-			wait := time.Duration(rand.Int31n(2000)) * time.Millisecond
-			fmt.Println(id, a.Name, "retrying in", wait)
-			time.Sleep(wait)
 		}
 
 		// Clean out the data dir if we are responsible for it before we
 		// try again, since the old ports may have gotten written to
 		// the data dir, such as in the Raft configuration.
 		if a.DataDir != "" {
 			if err := os.RemoveAll(a.DataDir); err != nil {
-				require.Fail("%s %s Error resetting data dir: %s", id, a.Name, err)
+				return fmt.Errorf("%s %s Error resetting data dir: %s", id, a.Name, err)
 			}
 		}
+
+		return fmt.Errorf("%s %s Error starting agent: %s", id, a.Name, err)
 	}
 
+	a.Agent = agent
+
 	// Start the anti-entropy syncer
 	a.Agent.StartSync()
 
+	err = a.waitForUp()
```

Comment:
Nitpick: could be the same style as above `if err = a.waitForUp(); err != nil {`.